// sms-ir/lib/v2/verificationCode.js
const {
    LOGGER
} = require('mango-logger');
const axios = require('axios');

class VerificationCode {

    async send(token, phoneNumber, code) {
        const result = await axios.post('http://RestfulSms.com/api/VerificationCode', {
            Code: code,
            MobileNumber: phoneNumber
        }, {
            headers: {
                'x-sms-ir-secure-token': token
            }
        }).then(e => {
            LOGGER.objToFile('sms-ir', 'VerificationCode', e.data);
            return e.data;
        }).catch(e => {
            LOGGER.errorToFile('sms-ir', e);
            throw new Error(e)
        })
        return result;
    }

    async sendTemplate(token, phoneNumber, code, templateId) {
        const result = await axios.post('http://RestfulSms.com/api/UltraFastSend', {
            ParameterArray: [{
                Parameter: "VerificationCode",
                ParameterValue: code
            }],
            Mobile: phoneNumber,
            TemplateId: templateId
        }, {
            headers: {
                'x-sms-ir-secure-token': token
            }
        }).then(e => {
            LOGGER.objToFile('sms-ir', 'VerificationCode', e.data);
            return e.data;
        }).catch(e => {
            LOGGER.errorToFile('sms-ir', e);
            throw new Error(e)
        })
        return result;
    }
}

module.exports = VerificationCode;
